test/metabase/query_processor/test_util.clj

```diff
@@ -487,18 +490,44 @@
                                [(:name table) (:name field)]))))
                    (t2/select-pks-set :model/Field :table_id [:in table-ids]))}))
 
+(defn- query-results [query]
+  (let [results (qp/process-query query)]
+    (or (get-in results [:data :results_metadata :columns])
+        (throw (ex-info "Missing [:data :results_metadata :columns] from query results" results)))))
+
+;;; TODO -- we should mark this deprecated, I just don't want to have to update a million usages.
 (defn card-with-source-metadata-for-query
   "Given an MBQL `query`, return the relevant keys for creating a Card with that query and matching `:result_metadata`.
 
     (t2.with-temp/with-temp [Card card (qp.test-util/card-with-source-metadata-for-query
                                         (data/mbql-query venues {:aggregation [[:count]]}))]
-      ...)"
+      ...)
+
+  Prefer [[metadata-provider-with-card-with-metadata-for-query]] instead of using this going forward."
   [query]
-  (let [results  (qp/process-userland-query query)
-        metadata (or (get-in results [:data :results_metadata :columns])
-                     (throw (ex-info "Missing [:data :results_metadata :columns] from query results" results)))]
-    {:dataset_query   query
-     :result_metadata metadata}))
+  {:dataset_query   query
+   :result_metadata (query-results query)})
+
+(defn metadata-provider-with-card-with-query-and-actual-result-metadata
```

Comment:
is it better if it supports a list of queries?

I think tests like `mlv2-references-in-join-conditions-test` will benefit from this.